### PR TITLE
drivers: power_domain: gpio: Allow enabled at boot

### PR DESCRIPTION
**Pull request by rtalbott-tmo**
Allow for gpio power domain to be enabled at boot time.

### CODE REVIEW

File: drivers/power_domain/power_domain_gpio.c [hunk 1]
```diff
@@ -116,6 +118,8 @@ static int pd_gpio_init(const struct device *dev)
 		/* Device is unpowered */
```
Comment (JordanYates) [inline drivers/power_domain/power_domain_gpio.c:118]: If regulators can now be powered on at boot the assumption here that a device is unpowered if it is on a power domain itself is no longer valid

File: dts/bindings/power-domain/power-domain-gpio.yaml [hunk 2]
```diff
@@ -28,3 +28,7 @@ properties:
     type: int
     default: 0
     description: Off delay time, in microseconds
+
+  boot-on:
+    type: boolean
+    description: Sets whether the domain to power on at boot
```
Comment (JordanYates) [inline dts/bindings/power-domain/power-domain-gpio.yaml:34]: My preference is for something closer to "If set the regulator is automatically powered on at boot time". Minor suggestion.

**Review by JordanYates — CHANGES_REQUESTED** [review]

File: dts/bindings/power-domain/power-domain-gpio.yaml (hunk 2 above)
Comment (rtalbott-tmo) [reply dts/bindings/power-domain/power-domain-gpio.yaml]: sure, updated

**Review by JordanYates — CHANGES_REQUESTED** [review]

File: drivers/power_domain/power_domain_gpio.c [hunk 3]
```diff
@@ -113,9 +115,13 @@ static int pd_gpio_init(const struct device *dev)
 	data->next_boot = K_TIMEOUT_ABS_US(cfg->off_on_delay_us);
 
 	if (pm_device_on_power_domain(dev)) {
-		/* Device is unpowered */
-		pm_device_init_off(dev);
-		rc = gpio_pin_configure_dt(&cfg->enable, GPIO_DISCONNECTED);
+		if (cfg->boot_on) {
```
[the remark this answers is not in the document]
Comment (rtalbott-tmo) [reply drivers/power_domain/power_domain_gpio.c:118]: Trying to understand the logic here.. 
If the device is powered, then we call suspend and put the GPIO inactive,
but - if it's not currently powered and set to power on at boot (cfg->boot) then make the GPIO active,
else - the device is off so disconnect the GPIO.

	if (pm_device_is_powered(dev)) {
		pm_device_init_suspended(dev);
		rc = gpio_pin_configure_dt(&cfg->enable, GPIO_OUTPUT_INACTIVE);
	} else if (cfg->boot_on) {
		rc = gpio_pin_configure_dt(&cfg->enable, GPIO_OUTPUT_ACTIVE);
	} else {
		pm_device_init_off(dev);
		rc = gpio_pin_configure_dt(&cfg->enable, GPIO_DISCONNECTED);
	}

Does this make sense? Or is there something I'm missing?

[the remark this answers is not in the document]
Comment (JordanYates) [reply drivers/power_domain/power_domain_gpio.c]: The original call (`pm_device_on_power_domain`) is used to determine if the GPIO regulator is itself powered by another GPIO regulator. Additionally, the previous assumption was that power domains are always powered off at boot.

Taken together, the original reasoning was that if this current device is on a power domain, it is not currently powered, so attempting to enable it via a GPIO is not possible. The new chain should be:
```
if (!pm_device_is_powered) {
        /* Device is not powered (we cannot respect boot_on) */
  	pm_device_init_off(dev);
	rc = gpio_pin_configure_dt(&cfg->enable, GPIO_DISCONNECTED);
} else if (cfg->boot_on) {
	rc = gpio_pin_configure_dt(&cfg->enable, GPIO_OUTPUT_ACTIVE);
} else {
	pm_device_init_suspended(dev);
	rc = gpio_pin_configure_dt(&cfg->enable, GPIO_OUTPUT_INACTIVE);
}
```

[the remark this answers is not in the document]
Comment (rtalbott-tmo) [reply drivers/power_domain/power_domain_gpio.c]: Thanks for the explanation.  I have a much better understanding now.
I have updated accordingly.

**Review by ceolin — APPROVED** [review]

**Comment by JordanYates** [comment]
Thinking about this a bit further, I don't believe the feature should be solved this way, as the expected state of drivers after booting should be consistent. Adding a device specific flag to make this driver consistent is a hack that will need to be removed later anyway. In principle, the expected states are:

- No `pm_device_runtime`, has power applied: `ACTIVE`
- No `pm_device_runtime`, does not have power: `OFF`
- `pm_device_runtime`, has power applied: `SUSPENDED`
- `pm_device_runtime`, does not have power: `OFF`

I am in the process of upstreaming initialisation work that does this, which I think is a more appropriate solution.

**Comment by rtalbott-tmo** [comment]
Sounds good.  I'm looking forward to seeing the PR.

**Comment by JordanYates** [comment]
Alternative that I believe does what you want: #60410

**Pull request closed without merging**